spamc.go

```diff
@@ -89,23 +228,56 @@ func (c *Client) write(
 	w := bufio.NewWriter(buf)
 	tp := textproto.NewWriter(w)
 
-	err := tp.PrintfLine("%v SPAMC/%v", cmd, clientProtocolVersion)
-	if err != nil {
+	// Make sure to seek to start of message
+	if _, err := message.Seek(0, io.SeekStart); err != nil {
 		return err
 	}
 
-	// Make the sure message always ends in \r\n, if it doesn't SA will just
-	// keep reading from the connection and it will block until it timouts.
-	if !strings.HasSuffix(message, "\r\n") {
-		message += "\r\n"
+	size := int64(-1)
+	switch v := message.(type) {
+	case *strings.Reader:
+		size = v.Size()
+	case *bytes.Reader:
+		size = v.Size()
+	case *os.File:
+		stat, err := v.Stat()
+		if err != nil {
+			return err
+		}
+		size = stat.Size()
 	}
 
-	// Always add Content-length header.
-	err = tp.PrintfLine("Content-length: %v", len(message))
+	// make sure to delete a present content-length (and use it if
+	// we could not set size above
+	if _, ok := headers[HeaderContentLength]; ok {
+		if size == -1 {
```

Comment:
Doing this in one `if` is a bit neater I think :-)

    if _, ok := headers[HeaderContentLength]; ok && size == -1 {

Reply:
Ah yea correct will fix that, oversight ;)

Reply:
True .. nice catch will refactor it.